include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -465,69 +477,79 @@ struct __parallel_transform_scan_static_single_group_submitter<_Inclusive, _Elem
     }
 };
 
-template <typename _Size, typename _KernelName>
-struct __parallel_copy_if_single_group_submitter;
+template <typename _KernelName>
+struct __parallel_copy_if_single_group_functor;
 
-template <typename _Size, typename... _ScanKernelName>
-struct __parallel_copy_if_single_group_submitter<_Size, __internal::__optional_kernel_name<_ScanKernelName...>>
+template <typename... _ScanKernelName>
+struct __parallel_copy_if_single_group_functor<__internal::__optional_kernel_name<_ScanKernelName...>>
 {
-    template <typename _InRng, typename _OutRng, typename _UnaryOp, typename _Assign>
-    __future<sycl::event, __result_and_scratch_storage<_Size>>
-    operator()(sycl::queue& __q, _InRng&& __in_rng, _OutRng&& __out_rng, std::size_t __n, _UnaryOp __unary_op,
-               _Assign __assign, std::uint16_t __n_uniform, std::uint16_t __wg_size)
+    template <typename _InRng, typename _OutRng, typename _Size, typename _UnaryOp, typename _Assign>
+    std::array<_Size, 2>
+    operator()(sycl::queue& __q, _InRng&& __in_rng, _OutRng&& __out_rng, _Size __n, _Size __n_out,
+               _UnaryOp __unary_op, _Assign __assign, std::uint16_t __n_uniform, std::uint16_t __wg_size)
     {
         using _ValueType = std::uint16_t;
-
         // This type is used as a workaround for when an internal tuple is assigned to std::tuple, such as
         // with zip_iterator
-        using __tuple_type =
-            typename ::oneapi::dpl::__internal::__get_tuple_type<std::decay_t<decltype(__in_rng[0])>,
-                                                                 std::decay_t<decltype(__out_rng[0])>>::__type;
+        using __tuple_type = typename oneapi::dpl::__internal::__get_tuple_type<
+            std::decay_t<decltype(__in_rng[0])>, std::decay_t<decltype(__out_rng[0])>>::__type;
 
-        using __result_and_scratch_storage_t = __result_and_scratch_storage<_Size>;
-        __result_and_scratch_storage_t __result{__q, 0};
+        __result_storage<_Size> __result{__q, 2};
 
-        auto __event = __q.submit([&](sycl::handler& __hdl) {
+        __q.submit([&](sycl::handler& __hdl) {
             oneapi::dpl::__ranges::__require_access(__hdl, __in_rng, __out_rng);
 
             // Local memory is split into two parts. The first half stores the result of applying the
             // predicate on each element of the input range. The second half stores the index of the output
             // range to copy elements of the input range.
-            auto __lacc = __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>(std::size_t(__n_uniform) * 2), __hdl);
-            auto __res_acc =
-                __result.template __get_result_acc<sycl::access_mode::write>(__hdl, __dpl_sycl::__no_init{});
+            auto __lacc = __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>(std::size_t(__n_uniform) * 2 + 1), __hdl);
```

Comment:
This line also should be updated as more SLM is used:

https://github.com/uxlfoundation/oneDPL/blob/238d76ae8b9dd932c11dde1a587a97a06bfeba61/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h#L915